TACC/Core-CMS-Custom
# bug/WP-113: Registration form entities bugs

## Overview
Addresses some bugs on entities (+ catches potential errors on contacts)
…

## Related

- [WP-113](https://jira.tacc.utexas.edu/browse/WP-113)
…

## Changes
- When errors occur on writing contact and/or entity to database, now any associated records with said buggy entry will be deleted from database
- In addition to above, if an error occurs on an entity/contact creation in database, writing any further entities or contacts will be prevented
- Custom error message on entity ID fields, and reworked how requiring at least one of these fields is handled
- Added upper limit on NAIC and license codes to match type on these fields in database
…

## Testing
***Testing the error handling on faulty entities or contacts will be difficult to do, looking for ideas on this.***

1. Open http://localhost:8000/register/request-to-submit/
2. Fill out form up to "Number/Code" (FEIN, NAIC, License) section on entity block, then try to submit form.
3. Confirm an error reading "Please fill in at least one Number/Code." appears on one of the three boxes in this section. (see UI below)
4. Fill in at least one of the three boxes and try to submit again and confirm no error appears on any of the three inputs.
5. Remove the value(s) you just added to this section and resubmit form; confirm the same error in (3) appears again.
6. Add new entities blocks and test steps 3-5 again.
7. Leave one of these errors on at least one entity block, then remove offending entity blocks and confirm the error disappears on form submission.
8. On NAIC/License inputs, confirm you cannot submit the form with a number larger than `2,147,483,648`.

## UI
<img width="1439" alt="image" src="https://github.com/TACC/Core-CMS-Custom/assets/43251554/0e721a5a-44bc-4f26-87ae-e101f96f9613">

…
